File: src/main/java/gargoyle/netsquares/prop/simple/DoubleProperty.java
package gargoyle.netsquares.prop.simple;

import gargoyle.netsquares.prop.RWProperty;

import java.io.Serializable;
import java.util.Objects;

public final class DoubleProperty extends RWProperty<Double> implements Comparable<DoubleProperty>, Serializable {
    public DoubleProperty(Double value) {
        super(value);
    }

    public DoubleProperty() {
        this(0);
    }

    public DoubleProperty(double value) {
        super(value);
    }

    @Override
    public int compareTo(DoubleProperty o) {
        return Double.compare(get(), o.get());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof RWProperty)) return false;
        if (!super.equals(o)) return false;
        RWProperty<?> that = (RWProperty<?>) o;
        return Objects.equals(get(), that.get());
    }

    public double getValue() {
        return get();
    }

    public void setValue(double value) {
        set(value);
    }
}

File: src/main/kotlin/gargoyle/netsquares/prop/simple/DoubleProperty.kt
package gargoyle.netsquares.prop.simple

import gargoyle.netsquares.prop.RWProperty

class DoubleProperty(value: Double = 0.0) : RWProperty<Double>(value), Comparable<DoubleProperty> {

    override fun compareTo(other: DoubleProperty): Int {
        return (get()).compareTo(other.get())
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RWProperty<*>) return false
        if (!super.equals(other)) return false
        return get() == other.get()
    }

    var value: Double
        get() = get()
        set(value) {
            set(value)
        }

    override fun hashCode(): Int = get().hashCode()
}
